### DISCUSSION TITLE
Failed to initialize OpenGL loader for "example_sdl_opengl3"

### DISCUSSION THREAD

**Discussion by hsbgit** [ocornut/imgui#4653] [New Users Build/Link/Run issues ONLY!]
Good day,

I managed to build example "example_sdl_opengl3". I did not change any line of code or the makefile or whatever.

Nonetheless, when executing the program I get the following output:

> ./example_sdl_opengl3 
> Failed to initialize OpenGL loader!
> example_sdl_opengl3: ../../backends/imgui_impl_opengl3.cpp:285: void ImGui_ImplOpenGL3_NewFrame(): Assertion `bd != __null && "Did you call ImGui_ImplOpenGL3_Init()?"' failed.
> Aborted (core dumped)

I am using the following Ubuntu system as a **virtual box** machine, almost clean installation:

> Distributor ID:	Ubuntu
> Description:	Ubuntu 20.04.3 LTS
> Release:	20.04
> Codename:	focal

What might be the problem here?

-------------

A related but different question: Am I right that I have to use the following files in my project if I want to work with SDL2?

- imgui_impl_sdl.h
- imgui_impl_sdl.cpp
- imgui_impl_opengl3_loader.h
- imgui_impl_opengl3.cpp
- imgui_impl_opengl3.h

I expected that imgui_impl_sdl.h/cpp would be sufficient but as it seems this backend requires the opengl3 backend.

Thanks for helping

**Comment by PathogenDavid** [comment]
> > Failed to initialize OpenGL loader!

This is the important error here: OpenGL is failing to initialize. Virtual machines have notoriously bad support for accelerated graphics, and [VirtualBox is one of the worst](https://askubuntu.com/questions/858407/how-to-update-to-latest-opengl-version-on-virtualbox-ubuntu-linux-machine).

If you only plan to use Dear ImGui and not render anything else (IE: you aren't making a game or some other 3D application) you might be able to get away with using one of the [third-party software rasterizers](https://github.com/ocornut/imgui/wiki/Useful-Extensions#software-rendererrasterizer). You might also be able to use the [experimental SDL_renderer backend](https://github.com/ocornut/imgui/blob/13cdf2ff0e554d2097841e742a26180440c69173/backends/imgui_impl_sdlrenderer.cpp) since I believe SDL_renderer supports software rendering. (Do note this requires SDL 2.0.17 which has not been released yet, so you'll have to build SDL yourself.)

If you're making something which needs accelerated rendering (IE: a game), you should not use VirtualBox.

--------------

> I expected that imgui_impl_sdl.h/cpp would be sufficient but as it seems this backend requires the opengl3 backend.

Dear ImGui backends typically come in two halves: The platform backend and the rendering backend.

The platform backend is responsible for interacting with the OS to handle things like keyboard and mouse input and (in the case of the `docking` branch) creating OS windows.

The rendering backend is responsible for submitting draw commands to actually render things on the screen.

In this case, `imgui_impl_sdl` is the platform backend and `imgui_impl_opengl3` is the rendering backend.

**Reply by hsbgit** [reply]
Thanks for your fast response and your explanation about the backends.

With respect to the failed init of the OpenGL loader I cloned and installed the latest SDL2 and checked the example_sdl_sdlrenderer example. This example starts as expected (though it is just software-based rendering).

Nonetheless, my question is answered, thanks!

// Edit: Oh, and please keep the renderer backend as it is quite useful for such cases (virtual machines) :)
